File: src/utils/configSchema.js
const VALID_ROUTERS = new Set(['Muskingum', 'RapidMuskingum', 'UnitMuskingum'])
const LEGACY_KEY_MAP = {
  routing_params_file: 'params_file',
  initial_state_file: 'channel_state_init_file',
  final_state_file: 'channel_state_final_file',
  input_type: 'runoff_processing_mode',
  catchment_volumes_files: 'qlateral_files',
  runoff_type: 'grid_accumulation_type',
  runoff_depths_files: 'grid_runoff_files',
  weight_table_file: 'grid_weights_file',
}
const REMOVED_KEYS = new Set(['connectivity_file', 'var_catchment_volume'])
const LIST_KEYS = new Set(['discharge_files', 'qlateral_files', 'grid_runoff_files'])

function isBlank(value) {
  return value === null || value === undefined || value === '' || (Array.isArray(value) && value.length === 0)
}

function migrateConfigKeys(raw) {
  const next = { ...raw }

  for (const [legacyKey, currentKey] of Object.entries(LEGACY_KEY_MAP)) {
    if (!isBlank(next[legacyKey]) && isBlank(next[currentKey])) {
      next[currentKey] = next[legacyKey]
    }
    delete next[legacyKey]
  }

  for (const key of REMOVED_KEYS) {
    delete next[key]
  }

  for (const key of LIST_KEYS) {
    if (typeof next[key] === 'string' && next[key].trim()) {
      next[key] = [next[key]]
    }
  }

  return next
}

function inferRouterFromFields(config) {
  if (!config || typeof config !== 'object') return null
  if (config.uh_kernel_file) return 'UnitMuskingum'
  if (
    (Array.isArray(config.qlateral_files) && config.qlateral_files.length > 0) ||
    (Array.isArray(config.grid_runoff_files) && config.grid_runoff_files.length > 0) ||
    config.grid_weights_file
  ) {
    return 'RapidMuskingum'
  }
  return null
}

export function inferRouter(config, fallback = 'Muskingum') {
  const candidate = String(config?._router || fallback || '').trim()
  return VALID_ROUTERS.has(candidate) ? candidate : 'Muskingum'
}

export function normalizeLoadedConfig(raw) {
  if (!raw || typeof raw !== 'object' || Array.isArray(raw)) {
    return { _router: 'Muskingum' }
  }
  const next = migrateConfigKeys(raw)
  const inferredRouter = next._router || inferRouterFromFields(next) || 'Muskingum'
  return { ...next, _router: inferRouter(next, inferredRouter) }
}

export function stripEmptyValues(config) {
  const cleaned = {}
  for (const [key, val] of Object.entries(config || {})) {
    if (key.startsWith('_')) continue
    if (val === null || val === undefined || val === '') continue
    if (Array.isArray(val)) {
      const compact = val.filter(v => v !== null && v !== undefined && v !== '')
      if (!compact.length) continue
      cleaned[key] = compact
      continue
    }
    cleaned[key] = val
  }
  return cleaned
}

File: river_route_app/config_schema.py
from __future__ import annotations

from typing import Any

VALID_ROUTERS: frozenset[str] = frozenset({
    'Muskingum',
    'RapidMuskingum',
    'UnitMuskingum',
})
LEGACY_KEY_MAP: dict[str, str] = {
    'routing_params_file': 'params_file',
    'initial_state_file': 'channel_state_init_file',
    'final_state_file': 'channel_state_final_file',
    'input_type': 'runoff_processing_mode',
    'catchment_volumes_files': 'qlateral_files',
    'runoff_type': 'grid_accumulation_type',
    'runoff_depths_files': 'grid_runoff_files',
    'weight_table_file': 'grid_weights_file',
}
REMOVED_KEYS: frozenset[str] = frozenset({
    'connectivity_file',
    'var_catchment_volume',
})


def _is_blank(value: Any) -> bool:
    return value is None or value == '' or value == []


def clean_config(config: dict[str, Any] | None) -> dict[str, Any]:
    """Keep only concrete, current-schema values and migrate known legacy keys."""
    if not isinstance(config, dict):
        return {}

    migrated: dict[str, Any] = {}
    for key, val in config.items():
        if not isinstance(key, str) or key.startswith('_') or key in REMOVED_KEYS:
            continue
        target_key = LEGACY_KEY_MAP.get(key, key)
        if target_key in REMOVED_KEYS:
            continue
        if target_key not in migrated or _is_blank(migrated[target_key]):
            migrated[target_key] = val

    cleaned: dict[str, Any] = {}
    for key, val in migrated.items():
        if val is None or val == '':
            continue
        if isinstance(val, list):
            compact = [x for x in val if x not in (None, '')]
            if not compact:
                continue
            cleaned[key] = compact
            continue
        cleaned[key] = val
    return cleaned


def resolve_router_name(router_name: str | None, default: str = 'Muskingum') -> str:
    router = (router_name or '').strip()
    if router in VALID_ROUTERS:
        return router
    return default if default in VALID_ROUTERS else 'Muskingum'
